libensemble/gen_funcs/persistent_tasmanian.py

```diff
@@ -44,15 +173,149 @@ def sparse_grid_batched(H, persis_info, gen_specs, libE_info):
+    H0['sim_id'] = np.arange(init_pts.shape[0], dtype='int')
+    H0['priority'] = np.flip(H0['sim_id'])
+    tag, Work, calc_in = ps.send_recv(H0)
+
+    # Subsequent runs.
+    while tag not in [STOP_TAG, PERSIS_STOP]:
+
+        # Parse the points returned by the allocator.
+        num_completed += calc_in['x'].shape[0]
+        queued_pts, queued_ids, offset = get_state(queued_pts, queued_ids, offset, completed_points=calc_in['x'], tol=tol)
+
+        # Compute the next batch of points (if they exist).
+        new_pts = np.empty((0, num_dims), dtype='float')
+        refined_pts = np.empty((0, num_dims), dtype='float')
+        refined_ord = np.empty(0, dtype='int')
+        if grid.getNumLoaded() < 1000 or num_completed > 0.2 * grid.getNumLoaded():
+            # A copy is needed because the data in the calc_in arrays are not contiguous.
+            grid.loadConstructedPoint(np.copy(calc_in['x']), np.copy(calc_in['f']))
+            if 'tasmanian_checkpoint_file' in U:
+                grid.write(U['tasmanian_checkpoint_file'])
+            refined_pts = get_refined_points(grid)
+            # If the refined points are empty, then there is a stopping condition internal to the Tasmanian sparse grid that is
+            # being triggered by the loaded points.
+            if refined_pts.size == 0:
+                break
+            refined_ord = np.lexsort(np.rot90(refined_pts))
+            delete_ind = get_2D_duplicate_indices(refined_pts, queued_pts, x_ord=refined_ord, tol=tol)
+            new_pts = np.delete(refined_pts, delete_ind, axis=0)
+
+        if new_pts.shape[0] > 0:
+            # Update the state variables with the refined points and update the queue in the allocator.
+            num_completed = 0
+            queued_pts, queued_ids, offset = get_state(queued_pts, queued_ids, offset, new_points=new_pts, tol=tol)
+            H0 = get_H0(gen_specs, refined_pts, refined_ord, queued_pts, queued_ids, tol=tol)
+            tag, Work, calc_in = ps.send_recv(H0)
+        else:
+            tag, Work, calc_in = ps.recv()
+
+    return [], persis_info, FINISHED_PERSISTENT_GEN_TAG
+
+
+def get_sparse_grid_inputs(user_specs, sim_f, num_dims, num_vals=1, mode='batched'):
+    """
+    Helper function that generates the simulator, generator, and allocator specs as well as the persis_info dictionary to ensure
+    that they are compatible with the custom generators in this script.
+
+    INPUTS:
+        user_specs (dict)   : a dictionary of user specs that is needed in the generator specs.
+        sim_f      (func)   : a lambda function that takes in generator outputs (simulator inputs) and returns simulator outputs.
+        num_dims   (int)    : number of model inputs.
+        num_vals   (int)    : number of model outputs.
+        mode       (string) : can either be 'batched' or 'async'.
+    """
+
+    assert 'tasmanian_init' in user_specs
+    assert mode in ['batched', 'async']
+
+    sim_specs = {
+		'sim_f' : sim_f,
```

Comment:
Inconsistent alignment.

Reply:
Fixed in a new PR. I followed Jeff's suggestion below.